src/main/java/com/microsoft/graph/authentication/IAuthenticationProvider.java

```diff
@@ -0,0 +1,20 @@
+package com.microsoft.graph.authentication;
+
+import java.net.URL;
+import java.util.concurrent.CompletableFuture;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Authenticates requests to be sent to the API
+ */
+public interface IAuthenticationProvider {
+    /**
+     * Authenticates the request
+     *
+     * @param requestUrl the outgoing request URL
+     * @return a future with the token
+     */
+    @Nonnull
+    CompletableFuture<String> getAuthorizationTokenAsync(@Nonnull final URL requestUrl);
```

Comment:
Is it actually necessary returning a Future here? A null future and a null access token are being handled in the same way anyway.

Reply:
the authentication token work is potentially async (think user interaction with the service, http calls...). If the consumer using an http client that allows for it, we want them to be able to chain the futures and avoid blocking at multiple levels.

Reply:
Thanks for the explanation!